src/doc/rustc/src/platform-support/unknown-uefi.md

```diff
@@ -230,3 +230,71 @@ pub extern "C" fn main(_h: efi::Handle, st: *mut efi::SystemTable) -> efi::Statu
     efi::Status::SUCCESS
 }
 ```
+
+## Rust std for UEFI
+This section contains information on how to use std on UEFI.
+
+### Build std
+The building std part is pretty much the same as the official [docs](https://rustc-dev-guide.rust-lang.org/getting-started.html).
+The linker that should be used is `rust-lld`. Here is a sample `config.toml`:
+```toml
+[llvm]
+download-ci-llvm = false
+[rust]
+lld = true
+[target.x86_64-unknown-uefi]
+linker = "rust-lld"
+```
+Then just build using `x.py`:
+```sh
+./x.py build --target x86_64-unknown-uefi
+```
+
+### Std Requirements
+The current std has a few basic requirements to function:
+1. Memory Allocation Services (`EFI_BOOT_SERVICES.AllocatePool()` and
+   `EFI_BOOT_SERVICES.FreePool()`) are available.
+If the above requirement is satisfied, the Rust code will reach `main`.
+Now we will discuss what the different modules of std use in UEFI.
+
+### Implemented features
+#### alloc
+- Implemented using `EFI_BOOT_SERVICES.AllocatePool()` and `EFI_BOOT_SERVICES.FreePool()`.
+- Passes all the tests.
+- Some Quirks:
+  - Currently uses `EfiLoaderData` as the `EFI_ALLOCATE_POOL->PoolType`.
+#### cmath
+- Provided by compiler-builtins.
+#### env
+- Just some global consants.
+#### locks
+- Uses `unsupported/locks`.
+- They should work for a platform without threads according to docs.
+#### os_str
+- Uses WTF-8 from windows.
+
+## Example: Hello World With std
+The following code is a valid UEFI application showing stdio in UEFI. It also
+uses `alloc` type `OsString` and `Vec`.
+
+This example can be compiled as binary crate via `cargo` using the toolchain
+compiled from the above source (named custom):
+
+```sh
+cargo +custom build --target x86_64-unknown-uefi
+```
+
+```rust,ignore (platform-specific)
+use r_efi::efi;
+use std::os::uefi::ffi::OsStrExt;
+use std::{ffi::OsString, panic};
+
+pub fn main() {
+  let st = std::os::uefi::env::system_table().as_ptr() as *mut efi::SystemTable;
+  let mut s: Vec<u16> = OsString::from("Hello World!\n").encode_wide().collect();
+  s.push(0);
+  let r =
+      unsafe { ((*(*st).con_out).output_string)((*st).con_out, s.as_ptr() as *mut efi::Char16) };
```

Comment:
I know you pulled this example from a preexisting test, but this is challenging for me to read. The fact that I can't see the definition of the `con_out` or `output_string` field in libstd makes it much harder. I don't believe this is a suitable example to promote to users. This should be written to show more of the types involved and make it clearer that this is... extracting a function pointer and then calling with two arguments to it, I think?

...And implicitly borrowing `&[u16]` to call `as_ptr()` and then casting it to a `*mut u16` (if I am reading that type alias correctly) seems at best unnecessary, and at worst potentially UB: just write `s.as_mut_ptr()`

Reply:
In UEFI everything is rooted in the system-table, and there usually in the boot-services or runtime-services. Each of these is represented by a rather big memory structure with lots of callbacks. The `r-efi` crate is nothing more than a copy of the data-types from the UEFI specification into rust.

The approach taken here is obvious to anyone knowledgable in UEFI, but certainly not to anyone else. However, the alternative is to either copy random function pointers from magic offset, or to open-code the data-types with lots of unrelated definitions. I am not sure either of those would make this example more useful (I can see that I am biased, though).

The thing is, nearly every UEFI framework ends up open-coding the datatypes or using something like EDK2 / r-efi. Hence, this approach is very common to almost all UEFI developers. Maybe you can elaborate a bit on how to make this more useful to outsiders?

Reply:
I'm aware. I didn't come into this review unprepared. I consulted with other Rust-on-UEFI experts on this review before preparing a single remark.

My comment here is, straight up, that
```rust
 ((*(*ident_a).field_c).field_d)((*ident_a).field_c, ident_b.as_ptr() as *mut type_alias) };
```
is approximately the most difficult to read form of this code, regardless of whatever platform details are in play. It is not atypical for programmers to deliberately use a new identifier after a cast, to allow introducing additional type annotations if necessary, and allowing each step to have its own comments, like so:
```rust
    unsafe {
        let con_out: *mut simple_text_output::Protocol = (*st).con_out;
        let out_str: extern "win64" fn(_: *mut simple_text_output::Protocol, *mut u16) = (*con_out).output_string;
        out_str(con_out, s.as_mut_ptr());
    }
```

It's not a _huge_ improvement, but conciseness is not a virtue in a short example program, because it is supposed to be an _example_, that is, illustrative, and r-efi has 30 types and type aliases named "Protocol". If I am referencing this example, as a programmer, I want to first not go cross-eyed due to my dyscalculia rebelling at that many parentheses ("rainbow parentheses" extensions aren't going to be installed on this .md, yanno?), and secondly to be able to refer to the documentation in r-efi and immediately know which of those 30 "Protocol"s I should click on.

An increase in expertise does not guarantee a reduction in this desire. They may have already generated a layout in their head of "what is important about the UEFI spec" that does not match r-efi's layout (nor the UEFI spec's layout, for that matter). And more generally, expertise makes it more likely the reader is someone who, if they are not already intimately acquainted with the r-efi library specifically, is more cautious about reusing another programmer's type definitions over FFI, being fully aware that linking can allow a misdeclared type, that using a poorly-declared type is undefined behavior, and so they may have a desire to first check to make sure they are correct.

And I am _very_ acquainted with that nuance of cross-language FFI. I have wound up _many_ times reviewing a type or function, originally declared in C, comparing against its declaration in Rust, written either by human hands or bindgen, and have felt compelled to redeclare it, either to use a more type-safe form, or because I simply happened to know better about how the layout algorithms, parameter passing, and unsafe code in Rust, happen to work. So I would like this to be in a form where it's _easy_ for future readers to check the work, here, instead of having a choice between taking a crate's correctness for granted, or having to fumble around navigating the crate's layout.

If useful documentation is added to these type definitions in the r-efi crate, even just cross-links into the spec, it would also be useful for relative novices.

Reply:
I must also stress the Rust tagline has been, for a while now:
"A language empowering **everyone** to build reliable and efficient software."

So, please take that "everyone" as far as it can be reasonably taken. There are not many "outsiders" you can, in fact, reliably count on being "outsiders" here. This is not a matter of an "us" versus a "them".

Reply:
I thought of "outsider" as someone with a fresh look on things rather than being biased by living in the firmware C-code bubble for too long. But I don't mind dropping this analogy.

I think the example can certainly use temporaries as you showed!

Regarding your comments on `r-efi`: The aim of the project is to transpose the 2500 pages of the UEFI specification from C to rust following simple rules and avoiding exceptions wherever possible. Thus, anyone with prior knowledge in the spec (e.g., from coding in C) can deduce how the structures and types are represented in rust and thus, preferably, does not need to consult `r-efi` docs at all. This includes the name of the types as well as the memory layout.
Extracting the documentation alongside the code from the spec is a colossal task no-one has attempted so far. Given that it is already quite tough to ensure the copied types have no errors, I feel like copying the documentation would just introduce errors that would do more harm than good. I would gladly discuss suggestions in the `r-efi` project!

Or do you feel like `std` should carry its own copy of the types? There are certainly advantages to both approaches.

Reply:
> ...And implicitly borrowing &[u16] to call as_ptr() and then casting it to a *mut u16 (if I am reading that type alias correctly) seems at best unnecessary, and at worst potentially UB: just write s.as_mut_ptr()

Can you elaborate on the possible UB? My impression was that `*const` and `*mut` are interchangeable for UB reasons. The `r-efi` project has a [comment](https://github.com/r-efi/r-efi/blob/main/src/lib.rs#L73) on this. If the reasoning is wrong, we should certainly fix this!

Using `as_mut_ptr()` is certainly more elegant, but it requires a mutable slice. Using `as_ptr()` allows immutable slices. Since the spec says the argument to `output_string` shall not be mutated, I believe it is safe to cast to `*mut`.

Note that `r-efi` explicitly uses `*mut` over `*const`, since `const` in C does *NOT* imply that there is no other mutable pointer at the same time, and thus there is no proper way to represent the `const` from C in rust-FFI (as described in `r-efi` as linked above). Suggestions to improve this situation are highly welcome!

Reply:
> Can you elaborate on the possible UB? My impression was that *const and *mut are interchangeable for UB reasons. The r-efi project has a [comment](https://github.com/r-efi/r-efi/blob/main/src/lib.rs#L73) on this. If the reasoning is wrong, we should certainly fix this!

Ah, I had not expected that it was non-mutating.

So, I am going to doublecheck with a few of my colleagues (read: T-opsem) because this is imperfectly defined (people literally are working on a huge revision to the semantics of this very subject we are discussing), but my understanding is it depends on what you obtain them from. It is as if an implicit capability becomes embedded in the pointer. If you obtain `*const T` from `&T`, you get a `*const T` which has the "privileges" of `&T`. If you cast to `*mut T`, it still has not acquired the privilege to mutate that target! If you obtain `*mut T` from `&mut T`, you have something that has the "privileges" of `&mut T`. If you then cast it to `*const T`, it retains those privileges. I am going to double back over the reading and talk to people to verify I am not horribly wrong according to the latest understanding.

Since this is only an _argument to a function_, and it is clear the argument shall not be mutated... a rule which doesn't end when the function ends, correct? C could stash a pointer from that function, but it still isn't supposed to mutate it later, because that pointer was obtained from a reduced privilege!... I believe that defining this particular type as accepting `*const T` may actually be the better representation for the cases of specific function arguments, since `*mut T` will coerce to `*const T` but not vice versa.

Rust does not necessarily care about the _mere existence_ of pointers. We already know that if the pointer was thrown across FFI that others have witnessed it. What we care about is that, for `&mut T`, those pointers act as if they are "dead".

But for `&T`, the story is more complicated. We should remember that `&T` is allowed to be `UnsafeCell<U>`, and thus it is possible for `&T` to have "inner mutability". In the generic case, reasoning about the mutability of the type behind the reference is fully disabled, and it only can do so during monomorphized steps. So, I believe `&T` is more conformant with the C-like "const" understanding than one might expect. I do think some caution is warranted here, but I think the caution one should take shifts significantly between function arguments and type definitions. These are different animals due to...

...must I say that accursed word?...

...contravariance.

Reply:
Okay, I checked and yes, what I said was almost exactly the current proposed model. The actions through the pointer use the permissions of the parent reference, and C can save 1 million pointers. It only causes UB if it actually uses one of the pointers derived from Rust code in a way that would violate things. Mere existence does not seem to change anything.

Reply:
Right. But UEFI allows you to retain pointers to the value *and* mutate it (at least in one direction). The function-prototypes are used on both ends, the caller and the implementer. For the caller `*const` sounds like the right choice, since you now give a very strong guarantee to the function that you do *not* manipulate the value in parallel. Furthermore, the function documents how long it can retain the pointer and access it (as part of the API documentation). So `*const` is an excellent choice.
But imagine you are the implementer. If you pick `*const`, it will suggest that your caller gives you that guarantee, while UEFI does not mandate this in any way. This is because the caller might be C code, and they interpret `const char *` as an assumption that the implementation does not manipulate the value, but the caller certainly can. So as an implementer I would very much want the rust API to use `*mut` to signal that I really cannot assume the guarantees of `&T`.

There is downsides to both approaches, and this is almost exclusively a question of avoiding tripping users. I don't think this is a question of UB (which I think you agree on, if I read your comment correctly). We chose `*mut` everywhere, similar to `NonNull<T>` using `*mut` and not providing a `ConstNonNull<T>`.

Reply:
Well, I have used some temps as suggested to make it more readable. However, I decided to keep the cast to `*mut` for now.
```rust
unsafe {
    let con_out: *mut simple_text_output::Protocol = (*st).con_out;
    let output_string: extern "efiapi" fn(_: *mut simple_text_output::Protocol, *mut u16) = (*con_out).output_string;
    output_string(con_out, s.as_ptr() as *mut efi::Char16)
};
```

Reply:
> But imagine you are the implementer. If you pick `*const`, it will suggest that your caller gives you that guarantee, while UEFI does not mandate this in any way. This is because the caller might be C code, and they interpret `const char *`

...I am really not sure what you mean here. No Rust function can assume, after it is passed `&T`, that the caller will not manipulate it later, and `*const T` is ultimately not `const char*`, so such an attempt to infer back such a guarantee into C from a Rust function signature is invalid.

Reply:
Rust has no raw pointer that matches the semantics of `const T *` in C. Hence, for this reason among others, `r-efi` uses `*mut` for all pointers, unconditionally.

I am unsure what your question or suggestion is. Do you want me to elaborate why `r-efi` uses `*mut`? Do you want this PR to avoid `r-efi` because it uses `*mut`? Are you just pointing out that in this instance you would have preferred `*const`? Or am I missing something? I would love to move this forward, but I feel a bit lost.

Reply:
I just am perplexed as to the justifications given for using `*mut T` over `*const T`. But we can drop this sideshow as ultimately my concerns have been resolved, as the real concern was the readability and verification of whether or not it was going to be used in a UB way (answer: no, apparently?).